### PR TITLE
Mark System.Console APIs as unsupported on Android

### PR DESCRIPTION
**Pull request by MaximLipnin**
Annotated public System.Console APIs throwing PNSE on Android.

~~No changes added to ref-part b/c no warnings appeared (maybe I missed something?).~~

Part of #47911.

### CODE REVIEW

**Comment by stephentoub** [comment]
> No changes added to ref-part b/c no warnings appeared (maybe I missed something?).

You need to add these to the ref for them to be meaningful to anyone outside of System.Console.  The purpose of the attributes is to give warnings to consumers of the API that they might be calling something that doesn't work.

**Comment by MaximLipnin** [comment]
> > No changes added to ref-part b/c no warnings appeared (maybe I missed something?).
> 
> You need to add these to the ref for them to be meaningful to anyone outside of System.Console. The purpose of the attributes is to give warnings to consumers of the API that they might be calling something that doesn't work.

Yes, but as far as I remember there should be the analyzer warnings saying that one is missing the attribute annotations in the respective ref part. I didn't get such a warning so that's why I left the comment about ref part. Actually no issue to add it as well.

**Comment by carlossanlop** [comment]
> Yes, but as far as I remember there should be the analyzer warnings saying that one is missing the attribute annotations in the respective ref part. 

@stephentoub @buyaa-n is this a bug in the analyzer?

**Comment by buyaa-n** [comment]
> > Yes, but as far as I remember there should be the analyzer warnings saying that one is missing the attribute annotations in the respective ref part.
> 
> @stephentoub @buyaa-n is this a bug in the analyzer?

Not sure if that is a bug in ApiCompat analyzer CC @safern

**Comment by MaximLipnin** [comment]
Could anyone please give another review to it?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs
Comment (marek-safar) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:16]: Why is it not supported? There does not seem to be anything notworking?

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs
Comment (marek-safar) [inline src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs:20]: It should use IsAndroid check if we keep it as unsupported

File: src/libraries/tests.proj
Comment (marek-safar) [inline src/libraries/tests.proj:27]: It'd be better to fix the test

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs
Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs:20]: `IsAndroid` check is not available here b/c "OperatingSystem doesn't contain" it during the build. "#if TARGET_ANDROID" doesn't work as well.

Comment (marek-safar) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs:20]: What's the error you get? The API is available since 5.0 https://docs.microsoft.com/en-us/dotnet/api/system.operatingsystem.isandroid?view=net-5.0

Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs:20]: It's because `Microsoft.Extensions.Logging.Console` targets `netstandard2.0;net461`

Comment (marek-safar) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs:20]: Oh, so the test is actually not a real test app but custom build version of Microsoft.Extension.Logging.Console library? @eerhardt why are we doing it this way?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs
Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs]: Yeah, this annotation seems to be irrelevant, Alexander suggested moving the existing `!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)` check from the private `DoesConsoleSupportAnsi()` method to the public ctor.

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs
Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs]: I pushed another change, perhaps this pragma is no longer topical.

File: src/libraries/System.Console/ref/System.Console.cs
Comment (eerhardt) [inline src/libraries/System.Console/ref/System.Console.cs:15]: These files are generated by a tool. We can't make manual changes to it because they will get undone when someone regenerates the file.

See [https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md)

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/AddConsoleFormatterTests.cs]: I'm not sure I understand the question. This is a "trimming test" that Microsoft.Extensions.Logging.Console works in a trimmed app.

If Android doesn't support Console, we should just skip this test on Android. See 

https://github.com/dotnet/runtime/blob/6b64618b0b09e430a514d5b718bd46ce08f0b359/src/libraries/Microsoft.Extensions.Logging.Console/tests/TrimmingTests/Microsoft.Extensions.Logging.Console.TrimmingTests.proj#L6

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:50]: @buyaa-n @jeffhandley - here is another scenario where https://github.com/dotnet/runtime/issues/44922 would help.

Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:50]: We might as well remove the duplicate check in `DoesConsoleSupportAnsi()` below, since this the is the only caller of it. And then rename the method to indicate it is `Windows-only`.

```suggestion
            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || DoesWindowsConsoleSupportAnsi())
```

**Comment by eerhardt** [comment]
>         [UnsupportedOSPlatform("browser")]

Why isn't `CursorLeft` and `CursorTop` annotated for Android?

---
Refers to: src/libraries/System.Console/src/System/Console.cs:423 in 54c4907. [](commit_id = 54c4907b4b45d63bcb8527db09cbd717434e19c1, deletion_comment = False)

**Comment by eerhardt** [comment]
>         [UnsupportedOSPlatform("browser")]

Why is `OpenStandardInput()` annotated as unsupported, but `OpenStandardInput(int bufferSize)` isn't?

---
Refers to: src/libraries/System.Console/src/System/Console.cs:555 in 54c4907. [](commit_id = 54c4907b4b45d63bcb8527db09cbd717434e19c1, deletion_comment = False)

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs
Comment (buyaa-n) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs:50]: @MaximLipnin could you add `// TODO update when https://github.com/dotnet/runtime/issues/44922 implemented `

File: src/libraries/System.Console/ref/System.Console.cs
Comment (MaximLipnin) [reply src/libraries/System.Console/ref/System.Console.cs]: Fixed, thanks

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs
Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs]: Updated, thanks

Comment (MaximLipnin) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProvider.cs]: Updated

File: src/libraries/System.Console/src/System/Console.cs
Comment (eerhardt) [inline src/libraries/System.Console/src/System/Console.cs:557]: What about `SetIn`, `Read()` and `ReadLine()`? Those are all annotated as `[UnsupportedOSPlatform("browser")]`, should they also have `[UnsupportedOSPlatform("android")]` on them?

Comment (MaximLipnin) [reply src/libraries/System.Console/src/System/Console.cs:557]: Addressed, thanks

**Comment by MaximLipnin** [comment]
Please, give it another review round 😄

**Review by jeffhandley — APPROVED** [review]
Looks good to me, but please also wait for a re-review from one of the previous reviewers.

**Review by marek-safar — APPROVED** [review]

**Pull request merged**
